src/main/java/ttt/GameLoop.java

```diff
@@ -31,37 +31,54 @@ private void playGame() {
         while (!gameEngine.isOver()) {
             String input = getLocation();
             display.clearScreen();
-            if (isInvalidInput(input)) {
-                write(consoleBoard.createBoard(gameEngine.showBoard()));
-                display.invalidInput();
-            } else {
-                placeMark(input);
-            }
+            processInput(input);
         }
     }
 
+    private String getLocation() {
+        write(consoleBoard.createBoard(gameEngine.showBoard()));
+        display.displayTurn(gameEngine.currentMark());
+        display.promptForLocation();
+        return gameEngine.getCurrentPlayer().getLocation(input, gameEngine.showBoard());
```

Comment:
All the method chaining here looks a little suspicious. Perhaps it is a sign that the design could be simplified.